### PR TITLE
Refactor `preprocess_cls` to `preprocess`, add `Serializer`, add `DataPipelineState`

### LINKED ISSUE #191: Support more return types in predict
Currently, the predict function returns only categories.

As a user I want to set the return type in the predict function, so I can get a label or raw probabilities instead of just the category.

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Bit of a big one. Makes the following changes:
- Switch from `{pre/post}process_cls` to just handing around references to instantiated objects. This includes:
    - removing classmethods called e.g. `instantiate_preprocess` (now just handled in the `__init__` of the class)
    - moving instantiation dependencies to the preprocess classes (e.g. default image transforms are now on the `ImageClassification` preprocess rather than the datamodule
- Adds a `Serializer` class so that a datamodule is now made up of a `Preprocess`, `Postprocess`, and a `Serializer`. These are classes which control the conversion from model output to desired prediction format (e.g. labels, classes etc.). They are seperate from the `Postprocess` as the user should be able to change the `Serializer` without messing with the more fundamental stuff in the `Postprocess` (without which the traning would usually fail). This Fixes #191 .
- Adds a `DataPipelineState`. Children of the data pipeline need to be able to communicate, this expands the `PreprocessState` to be shared across preprocess, postprocess, and serializer. This enables e.g. `Labels` serializer just gets the labels from the `ClassificationState` registered in the `load_data` of the `ImageClassificationPreprocess`.
- Adds an example showing multi_label image classification. Uses a subset of the movie poster genre data here: https://www.cs.ccu.edu.tw/~wtchu/projects/MoviePoster/
- Adds option to pass the `data_fetcher` as an argument to `from_load_data_inputs`

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by kaushikb11** [comment]
@ethanwharris Could you please add some description about the PR.

**Comment by ethanwharris** [comment]
@kaushikb11 will do, only a draft at the moment :smiley:

File: docs/source/reference/multi_label_classification.rst [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    class CustomViz(BaseVisualization):
+
+        def show_per_batch_transform(self, batch: Any, _):
+            images = batch[0]
+            image = make_grid(images, nrow=2)
+            image = T.to_pil_image(image, 'RGB')
+            image.show()
+
+
+    class CustomMultiLabelPreprocess(ImageClassificationPreprocess):
```
Comment (tchaton) [inline docs/source/reference/multi_label_classification.rst:59]: Why it this needed ?

File: docs/source/general/data.rst [hunk 2]
```diff
@@ -438,7 +454,9 @@ Example::
 
         samples = uncollate(batch)
 
-        return [per_sample_transform(sample) for sample in samples]
+        samples = [per_sample_transform(sample) for sample in samples]
+
+        return [serialize(sample) for sample in samples]
```
Comment (tchaton) [inline docs/source/general/data.rst:459]:
```suggestion
        # only if serializers are enabled.
        return [serialize(sample) for sample in samples]
```

File: flash/vision/detection/data.py [hunk 3]
```diff
@@ -208,12 +193,14 @@ def from_coco(
...
         batch_size: int = 4,
         num_workers: Optional[int] = None,
-        preprocess_cls: Type[Preprocess] = None,
+        preprocess: Preprocess = None,
         **kwargs
     ):
-
-        preprocess = cls.instantiate_preprocess(
-            train_transform, val_transform, predict_transform, predict_transform, preprocess_cls=preprocess_cls
+        preprocess = preprocess or ObjectDetectionPreprocess(
```
Comment (tchaton) [inline flash/vision/detection/data.py:199]: Should we still use `cls.preprocess_cls` there ?

Comment (tchaton) [reply flash/vision/detection/data.py:199]: A user could do:

```
class CustomObjectDetectionData(ObjectDetectionData):

     preprocess_cls = CustomObjectDetectionPreprocess

```

File: docs/source/reference/multi_label_classification.rst (hunk 1 above)
Comment (tchaton) [reply docs/source/reference/multi_label_classification.rst:59]: The class is created in fine-tuning example and doesn't exist in the predict one. Needs to be added.

File: flash/data/process.py [hunk 4]
```diff
@@ -483,3 +514,66 @@ def _save_data(self, data: Any) -> None:
...
+
+        Args:
+            sample: The output from the :class:`.Postprocess`.
+
+        Returns:
+            The serialized output.
+        """
+        return sample
+
+    def __call__(self, sample: Any):
```
Comment (edgarriba) [inline flash/data/process.py:546]:
```suggestion
    def __call__(self, sample: Any) -> Any:
```

**Review by edgarriba — COMMENTED** [review]
neat tests :)

File: flash/core/model.py [hunk 5]
```diff
@@ -84,8 +84,11 @@ def __init__(
         # TODO: should we save more? Bug on some regarding yaml if we save metrics
         self.save_hyperparameters("learning_rate", "optimizer")
 
-        self._preprocess = preprocess
-        self._postprocess = postprocess
+        self._preprocess: Optional[Preprocess] = preprocess
+        self._postprocess: Optional[Postprocess] = postprocess
+        self._serializer: Optional[Serializer] = None
+
+        self.serializer = serializer
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/core/model.py:91]: This is just to get it to call the setter, have added a comment

File: flash/core/classification.py [hunk 6]
```diff
@@ -11,36 +11,124 @@
...
+    Args:
+        labels: A list of labels, assumed to map the class index to the label for that class. If ``labels`` is not
+            provided, will attempt to get them from the :class:`.ClassificationState`.
+    """
+
+    def __init__(self, labels: Optional[List[str]] = None, multi_label: bool = False):
+        super().__init__(multi_label=multi_label)
+        self._labels = labels
+
+    def serialize(self, sample: Any) -> Union[int, List[int], str, List[str]]:
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/core/classification.py:114]: Not sure that's possible, the output type only depends on the size of the tensor

File: flash/core/classification.py [hunk 7]
```diff
@@ -11,36 +11,124 @@
...
+class Classes(ClassificationSerializer):
+    """A :class:`.Serializer` which applies an argmax to the model outputs (either logits or probabilities) and
+    converts to a list."""
+
+    def __init__(self, multi_label: bool = False, threshold: float = 0.5):
+        super().__init__(multi_label)
+
+        self.threshold = threshold
+
+    def serialize(self, sample: Any) -> Union[int, List[int]]:
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/core/classification.py:90]: `.tolist` will return an int if the input is a singleton tensor

File: docs/source/reference/multi_label_classification.rst [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+            dataset.num_classes = len(self.genres)
+            metadata = pd.read_csv(os.path.join(data, "metadata.csv"))
+
+            images = []
+            labels = []
+            for _, row in metadata.iterrows():
+                images.append(os.path.join(data, row['Id'] + ".jpg"))
+                labels.append(torch.IntTensor([row[genre] for genre in self.genres]))
+
+            return list(zip(images, labels))
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply docs/source/reference/multi_label_classification.rst]: Yeah, that's better, done :smiley:

File: docs/source/reference/multi_label_classification.rst [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    from flash.data.base_viz import BaseVisualization
+    from flash.data.utils import download_data
+    from flash.vision import ImageClassificationData, ImageClassifier
+    from flash.vision.classification.data import ImageClassificationPreprocess
+
+    # 1. Download the data
+    download_data("https://pl-flash-data.s3.amazonaws.com/movie_posters.zip", "data/")
+
+    # 2. Define our custom visualisation and datamodule
+    class CustomViz(BaseVisualization):
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply docs/source/reference/multi_label_classification.rst:49]: Yeah, can maybe switch later on after #228 is merged, although it's nice to show how a custom vis can be created

File: docs/source/reference/multi_label_classification.rst [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        "data/movie_posters/val/tt0361500.jpg",
+        "data/movie_posters/val/tt0361748.jpg",
+        "data/movie_posters/val/tt0362478.jpg",
+    ])
+    print(predictions)
+
+    # 4b. Or generate predictions with a whole folder!
+    datamodule = ImageClassificationData.from_folders(
+        predict_folder="data/movie_posters/predict/",
+        data_fetcher=CustomViz(),
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply docs/source/reference/multi_label_classification.rst:74]: Not sure what you mean? This bit is showing both prediction and visualisation at once so needs a datamodule.

**Review by tchaton — APPROVED** [review]
Amazing PR :)

**Review by edgarriba — APPROVED** [review]
